src/Agent.Worker/Release/ReleaseFileSystemManager.cs

```diff
@@ -0,0 +1,116 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+using Microsoft.VisualStudio.Services.Agent.Util;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Release
+{
+    [ServiceLocator(Default = typeof(ReleaseFileSystemManager))]
+    public interface IReleaseFileSystemManager : IAgentService
+    {
+        IEnumerable<FileInfo> GetFiles(string directoryPath, SearchOption searchOption);
+
+        StreamReader GetFileReader(string filePath);
+
+        Task WriteStreamToFile(Stream stream, string filePath);
+
+        void DeleteDirectory(string directoryPath);
+
+        void CleanupDirectory(string directoryPath);
+    }
+
+    public class ReleaseFileSystemManager : AgentService, IReleaseFileSystemManager
+    {
+        private RetryExecutor _retryExecutor = new RetryExecutor();
+        private const int StreamBufferSize = 1024;
+
+        public IEnumerable<FileInfo> GetFiles(string directoryPath, SearchOption searchOption)
+        {
+            return
+                Directory.GetFiles(ValidatePath(directoryPath), "*", searchOption)
+                    .Select(fullPath => new FileInfo(fullPath));
+        }
+
+        public IEnumerable<string> GetDirectories(string directoryPath, SearchOption searchOption)
+        {
+            return Directory.GetDirectories(ValidatePath(directoryPath), "*", searchOption);
+        }
+
+        public bool FileExists(string filePath)
+        {
+            return File.Exists(filePath);
+        }
+
+        public void DeleteFile(string filePath)
+        {
+            _retryExecutor.Execute(File.Delete, filePath);
+        }
+
+        public void DeleteDirectory(string directoryPath)
+        {
+            _retryExecutor.Execute(path => { Directory.Delete(path, true); }, directoryPath);
```

Comment:
There's a directory delete helper in IOUtil that handle the readonly attribute issue and also runs parallel... which per my understanding can increase the performance significantly (factor of 2 or 4 or something, i forgot, i saw a comment in filespec in tfs code that talked about it). If you have big directories to delete, you might want to use the IOUtil instead.